spec/models/order_spec.rb

```diff
@@ -77,6 +78,34 @@
     it{ is_expected.to have_db_column(:staff_note).of_type(:string)}
   end
 
+  describe '.counts_for' do
```

Comment:
WARNING: TDD MONOLOGUE INCOMING 😆  (just saw an opportunity to share, the code is actually good)

I love seeing specs ! :D

However, that is actually much more important to have tests on the **controllers** than models. Here's why : 

The code that counts orders could in theory be written on both the model level and the controller level.
However --> The controller is the **entry point** of the application. Meaning that if you test it there, you're actually testing that your server returns the expected value, and therefore behaves in the expected way.

Ideally it would also have an explicit name there, e.g `context 'counting orders by state'`

An example : Imagine the controller does not call `.counts_for` , then technically our server **isn't complying** with the specifications --> even though this test passes.
Ideally it is tested on both the controller and the model, but if you have to choose, the controller is where it should be tested (a more BDD approach).

When it comes to those end to end tests, the tests should not be aware of _how_ the code is written, it should test that the app behaves correctly under a set of circumstance. Think of all the specs of the server, it should be like reading a list of capabilities, or a documentation.

You can find more information in my book `How I annoyed my team to death, by Patrick`

Hope you'll find something interesting here @abulsayyad123 @namrataukirde @Darshan-Chauhan @abhinav098 
Everyone send good vibes to @steveyken , he has to deal with my rambling every day

Reply:
lol, can you send me the ISBN, I'll order a copy for everyone from BookDepository :D